java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageMetadataResult.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.ipc.message;
+
+import java.nio.ByteBuffer;
+
+import org.apache.arrow.flatbuf.Message;
+
+/**
+ * Class to hold resulting Message metadata and buffer containing the serialized Flatbuffer
+ * message when reading messages from a ReadChannel. This handles Message metadata only and
+ * does not include the message body data, which should be subsequently read into an ArrowBuf.
+ */
+public class MessageMetadataResult {
+
+  /**
+   * Construct a container to hold a deserialized Message metadata, and buffer
+   * with the serialized Message as read from a ReadChannel.
+   *
+   * @param messageLength the length of the serialized Flatbuffer message in bytes
+   * @param messageBuffer contains the serialized Flatbuffer Message metadata
+   * @param message the deserialized Flatbuffer Message metadata description
+   */
+  MessageMetadataResult(int messageLength, ByteBuffer messageBuffer, Message message) {
+    this.messageLength = messageLength;
+    this.messageBuffer = messageBuffer;
+    this.message = message;
+  }
+
+  /**
+   * Returns status indicating if the MessageResult has a valid message.
+   *
+   * @return true if the result contains a valid message
+   */
+  public boolean hasMessage() {
+    return message != null;
+  }
+
+  /**
+   * Get the length of the message in bytes.
+   *
+   * @return number of bytes in the message buffer.
+   */
+  public int getMessageLength() {
+    return messageLength;
+  }
+
+  /**
+   * Get the buffer containing the raw message bytes.
+   *
+   * @return buffer containing the message
+   */
+  public ByteBuffer getMessageBuffer() {
+    return messageBuffer;
+  }
+
+  /**
+   * Check if the message is valid and is followed by a body.
+   *
+   * @return true if message has a body
+   */
+  public boolean messageHasBody() {
```

Comment:
nit: Why is this not `hasMessageBody`?

Reply:
I think this goes with the idea that the body is separate from the message, rather than part of it. Either name seems fine to me

Reply:
Yeah, I didn't want it to sound like this class also contained the body. I'll make the description a little better here.

Reply:
I see. There is another function `getMessageBodyLength`. By using looking at the function name `messageHasBody` and `getMessageBodyLength` the API naming feels not too consistent. Does it make sense to rename `getMessageBodyLength` to `messageGetBodyLength` then?

Reply:
Hmm, I don't know. `Message` has a field `bodyLength` so I think `getMessageBodyLength()` is appropriate. Maybe if it's too confusing we should just get rid of `messageHasBody`?

Reply:
Yeah I think if this class contains just the metadata, it is a bit confusing to have this class contains methods for the body. User can always get such information from the `Message` object. 

By "When in doubt, leave it out" principle, I would prefer to get rid of the method too.